lmnet/lmnet/datasets/bdd100k.py

```diff
@@ -47,7 +65,7 @@ def count_max_boxes(cls, base_path=None):
     def num_max_boxes(self):
         # from dataset:
         # train - max 91 boxes, val - max 66 boxes
-        return self.max_boxes
```

Comment:
This is old code. I already updated with hard coded max value `91` here.

Reply:
self.max_boxes and max_boxes are defined and are used elsewhere. It is looks odd for me: 

https://github.com/blue-oil/blueoil/blob/3bdb78dec90e16360e5acb5751277cc0b40f7e46/lmnet/lmnet/datasets/bdd100k.py#L90-L92

https://github.com/blue-oil/blueoil/blob/3bdb78dec90e16360e5acb5751277cc0b40f7e46/lmnet/lmnet/datasets/bdd100k.py#L145-L147

https://github.com/blue-oil/blueoil/blob/3bdb78dec90e16360e5acb5751277cc0b40f7e46/lmnet/lmnet/datasets/bdd100k.py#L77-L79

Reply:
I don't understand your confusion. It is very logical if you think about the history and the issues of the blueoil development. Let me explain it for you:

First of all, in the `__init__` function the argument `max_boxes` has default value `100`. This is defined in old implementation of `bdd100k` dataset file, if you like you can check the commit : cc0e69a51a2782d4cf9880296a697c11cf5ed27f.

Previously, by default the max boxes was not in many object detection dataset class and the property `num_max_boxes` which actually returns the maximum number of boxes in one image by using `count _max_boxes` function to set the limit of the `numpy` array dimension which results in an uniform data structure for the training and validation purposes, which must have a fixed dimension. 

Now `count_max_boxes` function loads the Dataset classes for both `train` and `validation` set to determine the maximum number of boxes. In normal case this is OK. But in case of very large dataset like BDD100K, where we had performance issue before it was a waste of time to load the whole dataset twice just to determine the `num_max_boxes` property. Hence following the example of `widerface` dataset implementation a new variable `self.max_boxes` has been created which have the default value of 100 to keep the convension and can be modified from outside during initialization of the dataset class. The true number of `self.max_boxes` is being calculated in new implementation `bdd100k` dataset in the initialization phase using the code block:
```
num_boxes = len(bbox)
            if num_boxes > 0:
                self.paths.append(img_paths[img_name])
                self.bboxs.append(bbox)
                if num_boxes > max_boxes:
                    max_boxes = num_boxes
```
in `_init_files_and_annotations` function.

This has 1 problem, as during training there are 2 dataset in play, `train` and `validation`. By counting the number with the above code block does not consider both dataset at once, as using `count_max_boxes` function dose. So to avoid having wrong dataset array dimension for  one of the dataset the variable `self.max_boxes` is hard coded to `91` which is the max number of boxes in the dataset found by the code block mentioned above. This was also counted in previous implementation of `bdd100k` dataset and commented in the `num_max_boxes` property as you can check in commit : cc0e69a51a2782d4cf9880296a697c11cf5ed27f. The previous implementation set the default value 100 for the `self.max_boxes` which is adequate for both train and validation set.

But as we're using the maximum allowed image size, in order to optimize the memory footprint I hard coded the value to `91` while keeping the code block which calculate actual number of boxes for convenient purposes.

As you know, now we don't have the performance issue with the `BDD100K` dataset the original `count_max_boxes` function can be used, but it will load both the `train` and `validation` dataset twice, it will cost memory in the beginning of the training and might claim a large amount of `swap` space depending on available RAM. To avoid this issue the hard coded value `91` is kept and also the function `count_max_boxes` for keep the convension as it is being used or can be used for user side implementation in future.

I hope now it would be clear to you.

Reply:
I knew that the maximum number between validation and train was different because you explained. Although there are conventions as you say, in the current implementation the default value of `max_boxes` and `self.max_boxes` are unused variables. I think that the default value of max_boxes and self.max_boxes should be removed and can be fixed at 91. On the other hand, count_max_boxes is defined in the parent class, so it is unnatural if not implemented.

Reply:
I knew that the maximum number between validation and train was different because you explained. I understand there are conventions. But in the current implementation, the default value of max_boxes and self.max_boxes are unused variables. It looks odd to me. I think that the default value of max_boxes and self.max_boxes should be removed and can be fixed at 91. On the other hand, count_max_boxes is defined in the parent class, so it is unnatural if not implemented.

Reply:
@tkng San:
I agree with you. I also think `count_max_boxes` function can be removed from the `ObjectDetectionBase` class defined in `lmnet/lmnet/datasets/base.py` file, so that this function can be ignored if needed.

I also agree with you that we should not try to resolve multiple problem within one pull request. This pull-request is getting too much changes.

As a side note, from your suggestion of using self.max_boxes assigned in `_init_file_and_annotations` function but as I explained in my previous comment, by this way two datasets, `train` and `validation` dataset will have different number of array dimension. E.g. for this BDD100K dataset `train` set has maximum boxes of 91 and `validation` dataset will have 66 max boxes. This will create discrepancy while feeding the data into the label `placeholder`.

I think we should hardcode the `self.max_boxes` per dataset basis, as this is a constant value per dataset.